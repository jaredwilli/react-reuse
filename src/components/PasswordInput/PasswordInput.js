import React from 'react';
import PropTypes from 'prop-types';
import ProgressBar from '../ProgressBar';
import EyeIcon from '../EyeIcon';
import TextInput from '../TextInput';

/** Password input with integrated label, quality tips, and show password toggle. */
class PasswordInput extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            showPassword: false
        };
    }

    toggleShowPassword = event => {
        this.setState(prevState => {
            return {
                showPassword: !prevState.showPassword
            };
        });
    }

    render() {
        const { htmlId, value, label, error, onChange, placeholder, maxLength, showVisibilityToggle, quality, ...props } = this.props;
        const { showPassword } = this.state;

        return (
            <TextInput
                htmlId={htmlId}
                type={showPassword ? 'text' : 'password'}
                placeholder={placeholder}
                value={value}
                label={label}
                maxLength={maxLength}
                error={error}
                onChange={onChange}
                required
                {...props}>
                {showVisibilityToggle &&
                    <a onClick={this.toggleShowPassword}
                        style={{ marginLeft: 5 }}>
                        <EyeIcon />
                    </a>
                }
                {
                    (value.length > 0 && quality) && <ProgressBar percent={quality} width={130} />
                }
            </TextInput>
        );
    }
}


PasswordInput.propTypes = {
    /** Unique HTML ID. Used for tying label to HTML input. Handy hook for automated testing. */
    htmlId: PropTypes.string.isRequired,

    /** Input name. Recommend setting this to match object's property so a single change handler can be used by convention.*/
    name: PropTypes.string.isRequired,

    /** Password value */
    value: PropTypes.any,

    /** Input label */
    label: PropTypes.string,

    /** Function called when password input value changes */
    onChange: PropTypes.func.isRequired,

    /** Max password length accepted */
    maxLength: PropTypes.number,

    /** Placeholder displayed when no password is entered */
    placeholder: PropTypes.string,

    /** Set to true to show the toggle for displaying the currently entered password */
    showVisibilityToggle: PropTypes.bool,

    /** Display password quality visually via ProgressBar, accepts a number between 0 and 100 */
    quality: PropTypes.number,

    /** Validation error to display */
    error: PropTypes.string
};

PasswordInput.defaultProps = {
    maxLength: 50,
    showVisibilityToggle: false,
    label: 'Password'
};

export default PasswordInput;
