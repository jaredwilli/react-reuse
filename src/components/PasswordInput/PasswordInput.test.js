import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import PasswordInput from './PasswordInput';

describe('PasswordInput', () => {
    it('hides password quality by default', () => {
        const tree = renderer.create(<PasswordInput
            htmlId="test"
            name="test"
            onChange={() => {}}
            value="somevalue123" />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('shows password quality when enabled and a password is entered', () => {
        const tree = renderer.create(<PasswordInput
            htmlId="test"
            name="test"
            onChange={() => {}}
            showQuality
            value="somevalue123" />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('hides password quality when enabled but no password is entered', () => {
          const tree = renderer.create(<PasswordInput
            htmlId="test"
            name="test"
            onChange={() => {}}
            showQuality
            value="" />).toJSON();
          expect(tree).toMatchSnapshot();
    });
});
