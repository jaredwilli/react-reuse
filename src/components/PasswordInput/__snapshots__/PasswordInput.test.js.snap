// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput hides password quality by default 1`] = `
<div
  style={
    Object {
      "marginBottom": 16,
    }
  }
>
  <label
    htmlFor="test"
    style={
      Object {
        "display": "block",
      }
    }
  >
    Password
     
    <span
      style={
        Object {
          "color": "red",
        }
      }
    >
       *
    </span>
  </label>
  <input
    id="test"
    maxLength={50}
    name="test"
    onChange={[Function]}
    placeholder={undefined}
    style={undefined}
    type="password"
    value="somevalue123"
  />
</div>
`;

exports[`PasswordInput hides password quality when enabled but no password is entered 1`] = `
<div
  style={
    Object {
      "marginBottom": 16,
    }
  }
>
  <label
    htmlFor="test"
    style={
      Object {
        "display": "block",
      }
    }
  >
    Password
     
    <span
      style={
        Object {
          "color": "red",
        }
      }
    >
       *
    </span>
  </label>
  <input
    id="test"
    maxLength={50}
    name="test"
    onChange={[Function]}
    placeholder={undefined}
    showQuality={true}
    style={undefined}
    type="password"
    value=""
  />
</div>
`;

exports[`PasswordInput shows password quality when enabled and a password is entered 1`] = `
<div
  style={
    Object {
      "marginBottom": 16,
    }
  }
>
  <label
    htmlFor="test"
    style={
      Object {
        "display": "block",
      }
    }
  >
    Password
     
    <span
      style={
        Object {
          "color": "red",
        }
      }
    >
       *
    </span>
  </label>
  <input
    id="test"
    maxLength={50}
    name="test"
    onChange={[Function]}
    placeholder={undefined}
    showQuality={true}
    style={undefined}
    type="password"
    value="somevalue123"
  />
</div>
`;
